dmr_map2genome
==============

Map DMR/MRs to predefined genomic regions such as TSS, TES, 5dist, etc.

Required Parameters
-------------------
- ``in_cluster_out_path``: File path that contains all clustered PWMs and uncertain PWMs. Example: `../data/out/sep8_run1`.
- ``in_bpb3_ranking_path``: File path for bpb3 ranking results. Example: `../data/out/sep8_run1/test_67snps/foreground/test_seq/rankings/`.
- ``in_fileString4clustered_pwm``: File string or file folder/name used/contained for clustered PWMs, such as `quality_assessed_out_seed3`.

Required
--------
- ``-inSDFile IN_SORTEDDMR_FILE`` or ``--in_sortedDMR_file IN_SORTEDDMR_FILE``: An exported file from DMR_analysis with BED format, such as the exported file from "dmr_combine_multChrs4rank," where the columns are chrom, start_position, end_position, DMR_ID, and probability values. Please note that this input file should be sorted by chromosome position before being intersected with predefined genomic regions such as TSS, gene, or TES, etc.
- ``-inGRFile IN_GENEREGION_FILE`` or ``--in_geneRegion_file IN_GENEREGION_FILE``: A list of gene region files that is generated by hmst_seq_analyzer, such as "list_region_files.txt," where the file name and path of genomic region files are listed. These predefined genomic region files will be intersected with the sorted bed file of MR/DMR. Please refer to the demo of "run_hmsq.sh" for how to generate these predefined genomic region/annotation files.
- ``-inOFolder IN_OUTFILE_FOLDER`` or ``--in_outFile_folder IN_OUTFILE_FOLDER``: Path of the output file folder for exporting intersected MR/DMR files with each genomic region.
- ``-inRefFile IN_REFFLAT_FILE`` or ``--in_refFlat_file IN_REFFLAT_FILE``: Path of the input refFlat reference file, such as an export file "hg19.refFlat_clean_sorted.bed" from "hmst_seq_Analyzer".

Optional Parameters
-------------------
- ``-minOlp`` or ``--in_minimum_overlap4bedtools``: Minimum overlap rate in bedtools intersection. The default is 1e-9 or 1 bp overlap, which is a float number between 0 and 1.

